### ISSUE TITLE
Show marker across all 3 metric visualisation on dashboards 

### ISSUE THREAD

**Issue by jyothishjose6190** [hypertrace/hypertrace-ui#1240]
Show marker across all 3 metric visualizations on dashboards when we are going through any of the metric visualizations. This will help us to correlate latency spike easily with the error or call rate spike and vice versa.

**Comment by jyothishjose6190** [comment]
@JBAhire @anandtiwary  @aaron-steinfeld  Please share your thoughts here?

**Comment by jyothishjose6190** [comment]
@aaron-steinfeld  @anandtiwary  @itssharmasandeep  Whenever the mouse changes on a cartesian chart, an observable will emit those changes. All the other cartesian charts will be listening to the observable. If any change occurs, a tooltip will be shown on that particular location. This is how I plan to implement it.

**Comment by JBAhire** [comment]
@anandtiwary , we need to close this as it's been in requirements for quite some time now. Can we please close this asap?

**Comment by aaron-steinfeld** [comment]
My thoughts - first propose and settle on what the dashboard JSON would look like. That usually looks like a shared parent that handles it, or a shared identifier that they can use to coordinate (e.g. `synchronizedTooltipId: 'service-dash'`) Given that we're trying to correlate things across containers, my suspicion is the latter makes more sense. Also, radar in the above screenshot wouldn't make sense to include, since the synchronization is based on time. Probably only need it for cartesians.

Then, at the component level, we can figure out how to go from there. With the shared identifier, that likely could operate basically like a topic name, and using a shared service to publish and subscribe to tooltip events.

Anyway, will let @anandtiwary and @itssharmasandeep take it from here :)

**Comment by anandtiwary** [comment]
@JBAhire In the current implementation, we already show data from all three widgets in a single tooltip. Please refer to this screenshot. 
![Screenshot 2021-11-11 at 11 17 50 PM](https://user-images.githubusercontent.com/52081890/141409216-fa3fa6a0-1955-436f-9813-0b68d911a210.png)

What you have in the mock is very time-consuming to implement and honestly we had explored it in the past before finalizing the current solution (as shown in the screenshot above). Like Aaron explained above, this ask would require us to track and show mouse movements on adjacent widgets. We can definitely come up with a solution, but before we go there, please let us know what you think about the current solution as presented in the screenshot.

**Comment by JBAhire** [comment]
@anandtiwary , yes I am aware of the current implementation. The use-case Razorpay wants to solve here is being able to visualize and correlate deviations across all three key metrics. With current implementation I can see number but I am not sure if this is higher than usual, lower than usual or just normal pattern. Having 3 correlated markers provides us that capability.

**Comment by anandtiwary** [comment]
Okay, @jyothishjose6190 please start an implementation proposal. Mainly how would you do it, what changes would be required in cartesian chart, widget... would we need a separate widget? And what would be an overall estimate for this change. We can start from there.

**Comment by jyothishjose6190** [comment]
@aaron-steinfeld @anandtiwary @itssharmasandeep  I have created a JSON for the new container component named cartesian-container-widget.  Cartesian chart listens for all mouse event( mouseenter and mouseleave). Currently we are calling showTooltip method on mouseneter and hiding it on mouseleave. Instead of this, we can emit on mouse event  with locationdata  using an interaction handler. The interaction handler will update the service observable. All cartesian charts will be listening to this observable. They will show/hide tooltip based on this emitted value 

`{
        type: 'cartesian-container-widget',
        layout: {
          type: 'custom-container-layout',
          'enable-style': false,
          'column-dimensions': [
            {
              type: 'dimension-model',
              dimension: 1,
              unit: 'FR'
            },
            {
              type: 'dimension-model',
              dimension: 0.3,
              unit: 'FR',
              'min-dimension': 300
            }
          ],
          'row-dimensions': [
            {
              type: 'dimension-model',
              dimension: 1,
              unit: 'FR'
            }
          ],
          'cell-spans': [
            {
              type: 'cell-span-model',
              'col-start': 0,
              'col-end': 1,
              'row-start': 0,
              'row-end': 1
            },
            {
              type: 'cell-span-model',
              'col-start': 1,
              'col-end': 2,
              'row-start': 0,
              'row-end': 1
            }
          ]
        },
        children: [
          {
            type: 'cartesian-widget',
            'color-palette': RED_COLOR_PALETTE,
            'selectable-interval': true,
            'legend-position': LegendPosition.TopLeft,
            bands: [
              {
                type: 'band',
                name: 'Average Baseline',
                'upper-bound-name': 'Average Upper Bound',
                'lower-bound-name': 'Average Lower Bound',
                data: {
                  type: 'entity-metric-timeseries-data-source',
                  metric: {
                    type: 'metric-timeseries-band',
                    metric: 'duration',
                    aggregation: MetricAggregationType.Average
                  }
                }
              }
            ],
            series: [
              {
                type: 'series',
                name: 'P50',
                'visualization-type': 'line',
                data: {
                  type: 'entity-metric-timeseries-data-source',
                  metric: {
                    type: 'metric-timeseries',
                    metric: 'duration',
                    aggregation: MetricAggregationType.P50
                  }
                }
              },
              {
                type: 'series',
                name: 'P95',
                'visualization-type': 'line',
                data: {
                  type: 'entity-metric-timeseries-data-source',
                  metric: {
                    type: 'metric-timeseries',
                    metric: 'duration',
                    aggregation: MetricAggregationType.P95
                  }
                }
              },
              {
                type: 'series',
                name: 'P99',
                'visualization-type': 'line',
                data: {
                  type: 'entity-metric-timeseries-data-source',
                  metric: {
                    type: 'metric-timeseries',
                    metric: 'duration',
                    aggregation: MetricAggregationType.P99
                  }
                }
              },
              {
                type: 'series',
                name: 'Average',
                'visualization-type': 'line',
                data: {
                  type: 'entity-metric-timeseries-data-source',
                  metric: {
                    type: 'metric-timeseries',
                    metric: 'duration',
                    aggregation: MetricAggregationType.Average
                  }
                }
              }
            ],
            'selection-handler': {
              type: 'cartesian-explorer-selection-handler'
            },
            'hover-handler': {
              type: 'cartesian-mouse-hover-handler'
            }
          },
          {
            type: 'container-widget',
            layout: {
              type: 'custom-container-layout',
              'enable-style': false,
              'column-dimensions': [
                {
                  type: 'dimension-model',
                  dimension: 1,
                  unit: 'FR'
                }
              ],
              'row-dimensions': [
                {
                  type: 'dimension-model',
                  dimension: 1,
                  unit: 'FR'
                },
                {
                  type: 'dimension-model',
                  dimension: 1,
                  unit: 'FR'
                }
              ],
              'cell-spans': [
                {
                  type: 'cell-span-model',
                  'col-start': 0,
                  'col-end': 1,
                  'row-start': 0,
                  'row-end': 1
                },
                {
                  type: 'cell-span-model',
                  'col-start': 0,
                  'col-end': 1,
                  'row-start': 1,
                  'row-end': 2
                }
              ]
            },
            children: [
              {
                type: 'cartesian-widget',
                'selectable-interval': true,
                'legend-position': LegendPosition.None,
                'show-summary': true,
                title: 'Error Percentage',
                series: [
                  {
                    type: 'series',
                    name: 'Error Percentage',
                    color: '#de3618',
                    'visualization-type': 'area',
                    data: {
                      type: 'entity-error-percentage-timeseries-data-source'
                    }
                  }
                ],
                'selection-handler': {
                  type: 'cartesian-explorer-selection-handler'
                }
              },
              {
                type: 'cartesian-widget',
                'selectable-interval': true,
                'legend-position': LegendPosition.None,
                'show-summary': true,
                title: 'Error Rate',
                bands: [
                  {
                    type: 'band',
                    name: 'Baseline',
                    'upper-bound-name': 'Upper Bound',
                    'lower-bound-name': 'Lower Bound',
                    data: {
                      type: 'entity-metric-timeseries-data-source',
                      metric: {
                        type: 'metric-timeseries-band',
                        metric: 'errorCount',
                        aggregation: MetricAggregationType.AvgrateSecond
                      }
                    }
                  }
                ],
                series: [
                  {
                    type: 'series',
                    name: 'Error Rate',
                    color: '#de3618',
                    'visualization-type': 'area',
                    data: {
                      type: 'entity-metric-timeseries-data-source',
                      metric: {
                        type: 'metric-timeseries',
                        metric: 'errorCount',
                        aggregation: MetricAggregationType.AvgrateSecond
                      }
                    }
                  }
                ],
                'selection-handler': {
                  type: 'cartesian-explorer-selection-handler'
                },
                
              }
            ]
          }
        ]
      }`

**Comment by anandtiwary** [comment]
@jyothishjose6190 i think this is a good proposal and i would just like to make following small modifications. 
1. Since we still have to show tooltip on the primary widget, we can't remove the default show/hide tooltip behavior. But like you said, we can introduce a new mousemove event and emit to the service, and others charts can listen to this mouse location data.
2. We won't gain much by exposing the mousemove data to an interaction handler model And cartesian widget. The service can just live at cartesian component level only. 
3. Like Aaron suggested, we can introduce a group id, which we can define as a cartesian widget model property and a boolean for enabling mouse sync with other widgets.
4. Widget will send these properties to chart component.
5. Chart component will use sync group id to register itself with the service.
6. If sync flag is enabled, primary chart component will emit mouse move data to the same service.
7. If sync flag is enabled, chart will read the mouse move data from service (that is, its own sync group).
8. For non primary charts, only show marker. But for primary widget, show marker + tooltip.
9. You may have to support marker move in d3 chart in a similar fashion
